api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto

```diff
@@ -183,44 +183,162 @@ message JwtHeader {
-message JwtAuthentication {
-  // List of JWT rules to valide.
-  repeated JwtRule rules = 1;
+// This message specifies a Jwt requirement.
+message RequiredProvider {
+  // Specify the issuer for the required JwtProvider.
+  string issuer = 1;
 
-  // If true, the request is allowed if JWT is missing or JWT verification fails.
-  // Default is false, a request without JWT or failed JWT verification is not allowed.
-  bool allow_missing_or_failed = 2;
+  // This audiences field overrides the one in JwtProvider if not empty.
+  repeated string audiences = 2;
+}
+
+// This message specifies a list of RequiredProvider.
+message RequiredProviderList {
+  // Specify a list of RequiredProvider.
+  repeated RequiredProvider list = 1;
+}
+
+// This message specifies a Jwt requirement for a specific Route condition.
+// Example 1:
+//
+// .. code-block:: yaml
+//
+//    - match:
+//         prefix: "/healthz"
+//
+// In above example, There is no "requires" field for /healthz prefix match,
+// it means that requrests matched the path prefix don't require jwt authentication.
+//
+// Example 2:
+//
+// .. code-block:: yaml
+//
+//    - match:
+//         prefix: "/"
+//      requires: { issuer: "iss-A", audiences: [audience1, audience2]}
+//
+// In above example, all requests (all requests matched "/" path prefix),
+// require jwt authentication from "iss-A" with audiences override defined.
+message RequirementRule {
+  // The route matching parameter. Only when the match is satisfied, the "requires" field will
+  // apply.
+  //
+  // For example: following match will match all requests.
+  //
+  // .. code-block:: yaml
+  //
+  //    match:
+  //       prefix: "/"
+  //
+  envoy.api.v2.route.RouteMatch match = 1 [(validate.rules).message.required = true];
 
-  // This field lists the patterns allowed to bypass JWT verification. This only applies when
-  // `allow_missing_or_failed_jwt` is false. Under this config, if a request doesn't have JWT, it
-  // will be rejected. But some requests still needed to be forwarded without JWT, such as OPTIONS
-  // for CORS and some health checking paths.
+  // Specify a Jwt Requirement. If the message is empty, Jwt is not required.
   //
-  // Examples: bypass all CORS options requests
+  // .. code-block:: yaml
+  //
+  //   requires: { issuer: "iss-A" }
+  //
+  // or:
+  //   requires_any: { list: [{ issuer: "iss-A" }, { issuer: "iss-B" }] }
+  //
+  // or:
+  //   requires_all: { list: [{ issuer: "iss-A" }, { issuer: "iss-B" }] }
+  //
+  oneof requires_type {
+    // Specify a single RequiredProvider.
+    RequiredProvider requires = 2;
+
+    // Specify list of RequiredProvider. Their results are OR-ed.
+    // If any one of them passes, the result is passed.
+    RequiredProviderList requires_any = 3;
+
+    // Specify list of RequiredProvider. Their results are AND-ed.
+    // All of them must pass, if one of them fails or missing, it fails.
+    RequiredProviderList requires_all = 4;
```

Comment:
Have you considered a tree of protos with AND/OR semantics? Similar to what we do for log filtering, see https://github.com/envoyproxy/envoy/blob/c15019e79c832d9f0a09468affaadabc4be3e115/api/envoy/config/filter/accesslog/v2/accesslog.proto#L136

Reply:
This seems like it could fit the bill. The only real change from where the PR is now is an additional oneof message wrapper, right?

```
- rules:

  # require A
  - match: { prefix: "/baz" }
    requires: { issuer: "iss-A" }

  # require A or B
  - match: { prefix: "/foo" }
    requires_any:
      requirements:
      - requires: { issuer: "iss-A" }
      - requires: { issuer: "iss-B" }
  
  # require A and B
  - match: { prefix: "/bar" }
    requires_all:
      requirements:
      - requires: { issuer: "iss-A" }
      - requires: { issuer: "iss-B" }

  # require A and (B or C)
  - match: { prefix: "/qux" }
    requires_all:
      requirements:
      - requires: { issuer: "iss-A" }
      - requires_any:
          requirements:
          - requires: { issuer: "iss-B" }
          - requires: { issuer: "iss-C" }
```

I have a slight preference to inline the oneof into both the `RequirementRule` and the new message type, that way we can avoid the extraneous nested level, but I suppose that's not very DRY so I can understand either way.

Reply:
Like the idea.  Update the config file to support it.
I